source/common/router/router.cc

```diff
@@ -591,13 +591,20 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
     return Http::FilterHeadersStatus::StopIteration;
   }
 
+  // Increment the attempt count from 0 to 1 at the first upstream request.
+  attempt_count_++;
+  callbacks_->streamInfo().setAttemptCount(attempt_count_);
+
+  // Finalize the request headers before the host selection.
+  route_entry_->finalizeRequestHeaders(headers, callbacks_->streamInfo(),
+                                       !config_->suppress_envoy_headers_);
```

Comment:
finalizeRequestHeaders first could avoid to expose the `getRequestHostValue` method and could avoid multiple host calculating.